File: docker_images/pythonv1/wrapper/swagger_server/module_glue.py
# Copyright (c) Microsoft. All rights reserved.
# Licensed under the MIT license. See LICENSE file in the project root for
# full license information.
from swagger_server.iothub_client import IoTHubModuleClient, IoTHubTransportProvider, IoTHubMessage, IoTHubMessageDispositionResult, DeviceMethodReturnValue
from swagger_server.models.connect_response import ConnectResponse
import threading
import json

def transport_name_to_type(transport_name):
  return {
    'mqtt': IoTHubTransportProvider.MQTT,
    'mqttws': IoTHubTransportProvider.MQTT_WS,
    'amqp': IoTHubTransportProvider.AMQP,
    'amqpws': IoTHubTransportProvider.AMQP_WS,
    'http': IoTHubTransportProvider.HTTP
  }[transport_name]

class TwinCallbackContext:
  def __init__(self):
    self.event = threading.Event()
    self.latest_payload = {}
    self.current_complete_twin = {}

class MessageCallbackContext:
  def __init__(self):
    self.event = threading.Event()
    self.response = ""

class MethodInvokeContext:
  def __init__(self):
    self.event = threading.Event()
    self.payload = ""

class MethodCallbackContext:
  def __init__(self, expected_method_name, expected_request_payload, response, status_code):
    self.event = threading.Event()
    self.expected_method_name = expected_method_name
    self.actual_method_name = ""
    self.expected_request_payload = expected_request_payload
    self.actual_request_payload = ""
    self.response = response
    self.status_code = status_code

def patch_twin(prev_complete_twin, patch):
  j = prev_complete_twin
  if "desired" in patch:
    j["desired"] = patch["desired"]
  else:
    j["desired"] = patch
  return j

def twin_key(connection_id):
  return connection_id + "_twin"

def send_confirmation_callback(message, result, user_context):
  print('send confirmation callback complete ' + str(result))
  user_context.set()

def device_twin_callback(update_state, payload, user_context):
  print("device twin callback - state = {}".format(update_state))
  print(payload)
  payload = json.loads(payload)
  if (update_state == "COMPLETE"):
    print('device twin callback - COMPLETE')
    user_context.current_complete_twin = payload
  else:
    print('device twin callback - PARTIAL')
    user_context.latest_payload = payload
    user_context.current_complete_twin = patch_twin(user_context.current_complete_twin, user_context.latest_payload)
  user_context.event.set()

def reported_state_callback(status_code, user_context):
  print('reported state callback complete ' + str(status_code))
  user_context.set()

def receive_message_callback(message, user_context):
  print('received message callback')
  msg_str = message.get_string()
  if (msg_str):
    user_context.response = msg_str
  else:
    try:
      msg_byte_arr = message.get_bytearray()
    except:
      msg_byte_arr = "WARNING: get_bytearray threw exception"

    user_context.response = msg_byte_arr.decode("utf-8")
    user_context.event.set()
    return IoTHubMessageDispositionResult.ACCEPTED

def method_invoke_callback(response, user_context):
  print('method invoke callback complete')
  user_context.payload = response.responsePayload
  user_context.event.set()

def module_method_callback(method_name, payload, user_context):
  result = DeviceMethodReturnValue()
  print('module method callback triggered')
  user_context.actual_method_name = method_name
  user_context.actual_request_payload = json.loads(payload)

  if (user_context.actual_method_name == user_context.expected_method_name):
    if (user_context.actual_request_payload == user_context.expected_request_payload):
      print('method and payload matched. returning response')
      result.response = json.dumps(user_context.response)
      result.status = user_context.status_code
    else:
      print("request payload doesn't match")
      print("expected: " + user_context.expected_request_payload)
      print("received: " + user_context.actual_request_payload)
      result.status = 500
  else:
    print("method name doesn't match")
    print("expected: '" + user_context.expected_method_name + "'")
    print("received: '" + user_context.actual_method_name + "'")
    result.status = 404

  user_context.event.set()
  return result

class ModuleGlue:
  object_count = 1
  object_map = {}

  def _finish_connection(self, client):
    connection_id = 'moduleObject_' + str(self.object_count)
    self.object_count += 1
    self.object_map[connection_id] = client
    return ConnectResponse(connection_id)

  def connect_from_environment(self, transport_type):
    print ('connecting from environment')
    client = IoTHubModuleClient()
    IoTHubModuleClient.create_from_environment(client, transport_name_to_type(transport_type))
    return self._finish_connection(client)

  def connect(self, transport_type, connection_string, ca_certificate):
    print('connecting using ' + transport_type)
    client = IoTHubModuleClient(connection_string, transport_name_to_type(transport_type))
    if (ca_certificate.cert != None):
      client.set_option("TrustedCerts", ca_certificate.cert)
    return self._finish_connection(client)

  def disconnect(self, connection_id):
    print ('disconnecting ' + connection_id)
    if (connection_id in self.object_map):
      del self.object_map[connection_id]

  def enable_input_messages(self, connection_id):
    if (not connection_id in self.object_map):
      raise Exception("connection_id is invalid")

  def enable_methods(self, connection_id):
    if (not connection_id in self.object_map):
      raise Exception("connection_id is invalid")

  def enable_twin(self, connection_id):
    if (not connection_id in self.object_map):
      raise Exception("connection_id is invalid")
    client = self.object_map[connection_id]
    context = TwinCallbackContext()
    client.set_module_twin_callback(device_twin_callback, context)
    print('set module twin callback complete.  waiting')
    context.event.wait()
    context.event.clear()
    self.object_map[twin_key(connection_id)] = context

  def send_event(self, connection_id, event_body):
    print ('sending event on ' + connection_id)
    event = threading.Event()
    msg = IoTHubMessage(event_body)
    self.object_map[connection_id].send_event_async(msg, send_confirmation_callback, event)
    print('send complete.  waiting')
    event.wait()
    print('send confirmation received')

  def wait_for_input_message(self, connection_id, input_name):
    if (not connection_id in self.object_map):
      raise Exception("connection_id is invalid")
    client = self.object_map[connection_id]
    context = MessageCallbackContext()
    client.set_message_callback(input_name, receive_message_callback, context)
    print('waiting for input message')
    context.event.wait()
    print('input message received')

    #unset the callback
    #client.set_message_callback(input_name, None, None)
    return context.response

  def invoke_module_method(self, connection_id, device_id, module_id, method_invoke_parameters):
    if (not connection_id in self.object_map):
      raise Exception("connection_id is invalid")
    client = self.object_map[connection_id]
    method_name = method_invoke_parameters['methodName']
    payload = method_invoke_parameters['payload']
    timeout = method_invoke_parameters['responseTimeoutInSeconds']
    context = MethodInvokeContext()
    client.invoke_method_async(device_id, module_id, method_name, payload, timeout, method_invoke_callback, context)
    print('waiting for module method invoke response')
    context.event.wait()
    print('module method invoke response received')
    return context.payload

  def invoke_device_method(self, connection_id, device_id, method_invoke_parameters):
    if (not connection_id in self.object_map):
      raise Exception("connection_id is invalid")
    client = self.object_map[connection_id]
    method_name = method_invoke_parameters['methodName']
    payload = method_invoke_parameters['payload']
    timeout = method_invoke_parameters['responseTimeoutInSeconds']
    context = MethodInvokeContext()
    client.invoke_method_async(device_id, method_name, payload, timeout, method_invoke_callback, context)
    print('waiting for device method invoke response')
    context.event.wait()
    print('device method invoke response received')
    return context.payload

  def roundtrip_method_call(self, connection_id, methodName, requestAndResponse):
    if (not connection_id in self.object_map):
      raise Exception("connection_id is invalid")
    client = self.object_map[connection_id]
    context = MethodCallbackContext(methodName, requestAndResponse.request_payload['payload'], requestAndResponse.response_payload, requestAndResponse.status_code)
    client.set_module_method_callback(module_method_callback, context)
    print('waiting for method call')
    context.event.wait()
    print('method call received')

  def send_output_event(self, connection_id, output_name, event_body):
    print ('sending event on ' + connection_id + ',' + output_name)
    event = threading.Event()
    msg = IoTHubMessage(event_body)
    self.object_map[connection_id].send_event_async(output_name, msg, send_confirmation_callback, event)
    print('send complete.  waiting')
    event.wait()
    print('send confirmation received')

  def wait_for_desired_property_patch(self, connection_id):
    if (not connection_id in self.object_map):
      raise Exception("connection_id is invalid")
    if (not twin_key(connection_id) in self.object_map):
      raise Exception("no twin context")
    context = self.object_map[twin_key(connection_id)]
    print("waiting for twin patch response")
    context.event.wait()
    context.event.clear()
    print("twin patch response received")
    return context.latest_payload

  def get_twin(self, connection_id):
    if (not connection_id in self.object_map):
      raise Exception("connection_id is invalid")
    if (not twin_key(connection_id) in self.object_map):
      raise Exception("no twin context")
    context = self.object_map[twin_key(connection_id)]

    return { "properties": context.current_complete_twin }

  def send_twin_patch(self, connection_id, props):
    if (not connection_id in self.object_map):
      raise Exception("connection_id is invalid")
    client = self.object_map[connection_id]
    event = threading.Event()
    propstring = json.dumps(props)
    IoTHubModuleClient.send_reported_state(client, propstring, len(propstring), reported_state_callback, event)
    print("waiting for send reported state confirmation")
    event.wait()
    print("send reported state confirmation received")

  def cleanup_resources(self):
    listcopy  = list(self.object_map.keys())
    for key in listcopy:
      print("object {} not cleaned up".format(key))
      self.disconnect(key)

File: docker_images/java/wrapper/src/main/java/glue/ModuleGlue.java
package glue;

import com.microsoft.azure.sdk.iot.device.ClientOptions;
import com.microsoft.azure.sdk.iot.device.IotHubClientProtocol;
import com.microsoft.azure.sdk.iot.device.IotHubEventCallback;
import com.microsoft.azure.sdk.iot.device.IotHubMessageResult;
import com.microsoft.azure.sdk.iot.device.IotHubStatusCode;
import com.microsoft.azure.sdk.iot.device.Message;
import com.microsoft.azure.sdk.iot.device.ModuleClient;
import com.microsoft.azure.sdk.iot.device.auth.IotHubSSLContext;
import com.microsoft.azure.sdk.iot.device.edge.DirectMethodRequest;
import com.microsoft.azure.sdk.iot.device.exceptions.ModuleClientException;
import com.microsoft.azure.sdk.iot.device.twin.DirectMethodPayload;
import com.microsoft.azure.sdk.iot.device.twin.DirectMethodResponse;
import com.microsoft.azure.sdk.iot.device.twin.MethodCallback;
import com.microsoft.azure.sdk.iot.device.twin.Property;
import com.microsoft.azure.sdk.iot.device.twin.TwinCollection;
import io.swagger.server.api.MainApiException;
import io.swagger.server.api.model.Certificate;
import io.swagger.server.api.model.ConnectResponse;
import io.swagger.server.api.model.EventBody;
import io.swagger.server.api.model.MethodInvoke;
import io.swagger.server.api.model.MethodRequestAndResponse;
import io.swagger.server.api.model.Twin;
import io.vertx.core.AsyncResult;
import io.vertx.core.Future;
import io.vertx.core.Handler;
import io.vertx.core.json.Json;
import io.vertx.core.json.JsonObject;

import java.io.IOException;
import java.security.KeyManagementException;
import java.security.KeyStoreException;
import java.security.NoSuchAlgorithmException;
import java.security.cert.CertificateException;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.Map;
import java.util.Set;
import java.util.Timer;
import java.util.TimerTask;
import java.util.concurrent.TimeoutException;


public class ModuleGlue
{
    private IotHubClientProtocol transportFromString(String protocolStr)
    {
        IotHubClientProtocol protocol = null;

        if (protocolStr.equals("amqp"))
        {
            protocol = IotHubClientProtocol.AMQPS;
        }
        else if (protocolStr.equals("mqtt"))
        {
            protocol = IotHubClientProtocol.MQTT;
        }
        else if (protocolStr.equals("amqpws"))
        {
            protocol = IotHubClientProtocol.AMQPS_WS;
        }
        else if (protocolStr.equals("mqttws"))
        {
            protocol = IotHubClientProtocol.MQTT_WS;
        }
        else if (protocolStr.equals("http"))
        {
            protocol = IotHubClientProtocol.HTTPS;
        }
        return protocol;
    }

    HashMap<String, ModuleClient> _map = new HashMap<>();
    int _clientCount = 0;

    public void connectFromEnvironment(String transportType, Handler<AsyncResult<ConnectResponse>> handler)
    {
        System.out.printf("ConnectFromEnvironment called with transport %s%n", transportType);

        IotHubClientProtocol protocol = this.transportFromString(transportType);
        if (protocol == null)
        {
            handler.handle(Future.failedFuture(new MainApiException(500, "invalid transport")));
            return;
        }

        try
        {
            ModuleClient client = ModuleClient.createFromEnvironment(new UnixDomainSocketChannelImpl(), protocol);
            client.open(true);

            this._clientCount++;
            String connectionId = "moduleClient_" + this._clientCount;
            this._map.put(connectionId, client);

            ConnectResponse cr = new ConnectResponse();
            cr.setConnectionId(connectionId);
            handler.handle(Future.succeededFuture(cr));
        } catch (Exception e)
        {
            handler.handle(Future.failedFuture(e));
        }
    }

    private ModuleClient getClient(String connectionId)
    {
        return this._map.getOrDefault(connectionId, null);
    }

    public void connect(String transportType, String connectionString, Certificate caCertificate, Handler<AsyncResult<ConnectResponse>> handler)
    {
        System.out.printf("Connect called with transport %s%n", transportType);

        IotHubClientProtocol protocol = this.transportFromString(transportType);
        if (protocol == null)
        {
            handler.handle(Future.failedFuture(new MainApiException(500, "invalid transport")));
            return;
        }

        try
        {
            String cert = caCertificate.getCert();
            ClientOptions.ClientOptionsBuilder clientOptionsBuilder = ClientOptions.builder();
            if (cert != null && !cert.isEmpty())
            {
                clientOptionsBuilder.sslContext(IotHubSSLContext.getSSLContextFromFile(cert));
            }

            ModuleClient client = new ModuleClient(connectionString, protocol, clientOptionsBuilder.build());

            client.open(true);

            this._clientCount++;
            String connectionId = "moduleClient_" + this._clientCount;
            this._map.put(connectionId, client);

            ConnectResponse cr = new ConnectResponse();
            cr.setConnectionId(connectionId);
            handler.handle(Future.succeededFuture(cr));
        } catch (Exception e)
        {
            handler.handle(Future.failedFuture(e));
        }
    }


    public void invokeDeviceMethod(String connectionId, String deviceId, MethodInvoke methodInvokeParameters, Handler<AsyncResult<Object>> handler)
    {
        ModuleClient client = getClient(connectionId);
        if (client == null)
        {
            handler.handle(Future.failedFuture(new MainApiException(500, "invalid connection id")));
        }
        else
        {
            DirectMethodRequest request = new DirectMethodRequest(methodInvokeParameters.getMethodName(), Json.encode(methodInvokeParameters.getPayload()), methodInvokeParameters.getResponseTimeoutInSeconds(), methodInvokeParameters.getConnectTimeoutInSeconds());
            try
            {
                com.microsoft.azure.sdk.iot.device.edge.DirectMethodResponse result = client.invokeMethod(deviceId, request);
                handler.handle(Future.succeededFuture(makeMethodResultThatEncodesCorrectly(result)));
            } catch (Exception e)
            {
                handler.handle(Future.failedFuture(e));
            }
        }
    }

    private void _closeConnection(String connectionId)
    {
        System.out.printf("Disconnect for %s%n", connectionId);
        ModuleClient client = getClient(connectionId);
        if (client != null)
        {
            this._deviceTwinStatusCallback.setHandler(null);
            client.close();
            this._map.remove(connectionId);
        }
    }

    public void disconnect(String connectionId, Handler<AsyncResult<Void>> handler)
    {
        this._closeConnection(connectionId);
        handler.handle(Future.succeededFuture());
    }

    public void enableInputMessages(String connectionId, Handler<AsyncResult<Void>> handler)
    {
        ModuleClient client = getClient(connectionId);
        if (client == null)
        {
            handler.handle(Future.failedFuture(new MainApiException(500, "invalid connection id")));
        }
        else
        {
            handler.handle(Future.succeededFuture());
        }
    }

    private Twin _twin = null;
    private Handler<AsyncResult<Twin>> _twinHandler;
    private Timer _timer = null;

    public void setTwinHandler(Handler<AsyncResult<Twin>> handler)
    {
        if (handler == null)
        {
            this._twin = null;
        }
        else
        {
            this._twin = new Twin(new JsonObject(), new JsonObject());

        }
        this._twinHandler = handler;
    }

    public void onPropertyChanged(Property property, Object context)
    {
        System.out.println(
            "onProperty callback for " + (property.getIsReported() ? "reported" : "desired") +
                " property " + property.getKey() +
                " to " + property.getValue() +
                ", Properties version:" + property.getVersion());
        if (this._twin == null)
        {
            System.out.println("nobody is listening for desired properties.  ignoring.");
        }
        else
        {
            if (property.getIsReported())
            {
                ((JsonObject) this._twin.getReported()).getMap().put(property.getKey(), property.getValue());
            }
            else
            {
                ((JsonObject) this._twin.getDesired()).getMap().put(property.getKey(), property.getValue());
            }
            System.out.println(this._twin.toString());
            System.out.println("scheduling timer");
            this.rescheduleTwinHandler();
        }
    }

    private void rescheduleTwinHandler()
    {
        if (_twinHandler == null)
        {
            return;
        }
        // call _handler 2 seconds after the last designed property change
        if (this._timer != null)
        {
            this._timer.cancel();
            this._timer = null;
        }
        this._timer = new Timer();
        this._timer.schedule(new TimerTask()
        {
            @Override
            public void run()
            {
                _timer = null;
                if (_twinHandler != null && _twin != null)
                {
                    System.out.println("It's been 2 seconds since last desired property arrived.  Calling handler");
                    System.out.println(_twin.toString());
                    _twinHandler.handle(Future.succeededFuture(_twin));
                    _twinHandler = null;
                }
            }
        }, 2000);
    }

    private static class IotHubEventCallbackImpl implements IotHubEventCallback
    {
        Handler<AsyncResult<Void>> _handler = null;

        public void setHandler(Handler<AsyncResult<Void>> handler)
        {
            this._handler = handler;
        }

        @Override
        public void execute(IotHubStatusCode status, Object context)
        {
            System.out.println("IoT Hub responded to operation with status " + status.name());
            Handler<AsyncResult<Void>> handler = this._handler;
            this._handler = null;
            if (handler != null)
            {
                if (status == IotHubStatusCode.OK)
                {
                    handler.handle(Future.succeededFuture());
                }
                else
                {
                    handler.handle(Future.failedFuture(new MainApiException(500, "operation failed")));
                }
            }
        }
    }

    private IotHubEventCallbackImpl _deviceTwinStatusCallback = new IotHubEventCallbackImpl();

    public void enableTwin(String connectionId, final Handler<AsyncResult<Void>> handler)
    {
        final ModuleClient client = getClient(connectionId);
        if (client == null)
        {
            handler.handle(Future.failedFuture(new MainApiException(500, "invalid connection id")));
        }
        else
        {
            try
            {
                System.out.println("calling subscribeToDesiredProperties to start twin");
                client.subscribeToDesiredProperties(
                    (twin, context) ->
                    {
                        TwinCollection desiredProperties = twin.getDesiredProperties();
                        for (String key : desiredProperties.keySet())
                        {
                            onPropertyChanged(new Property(key, desiredProperties.get(key)), null);
                        }
                    },
                    null);

                handler.handle(Future.succeededFuture());
            }
            catch (Exception e)
            {
                handler.handle(Future.failedFuture((e)));
            }
        }
    }

    private static class EventCallback implements IotHubEventCallback
    {
        Handler<AsyncResult<Void>> _handler;

        public EventCallback(Handler<AsyncResult<Void>> handler)
        {
            this._handler = handler;
        }

        public synchronized void execute(IotHubStatusCode status, Object context)
        {
            System.out.printf("EventCallback called with status %s%n", status.toString());
            this._handler.handle(Future.succeededFuture());
        }
    }

    private void sendEventHelper(String connectionId, Message msg, Handler<AsyncResult<Void>> handler)
    {
        System.out.println("inside sendEventHelper");

        ModuleClient client = getClient(connectionId);
        if (client == null)
        {
            handler.handle(Future.failedFuture(new MainApiException(500, "invalid connection id")));
        }
        else
        {
            EventCallback callback = new EventCallback(handler);
            System.out.printf("calling sendEventAsync%n");
            client.sendEventAsync(msg, callback, null);
        }
    }

    public void sendEvent(String connectionId, EventBody eventBody, Handler<AsyncResult<Void>> handler)
    {
        System.out.printf("moduleConnectionIdEventPut called for %s%n", connectionId);
        System.out.println(eventBody);
        this.sendEventHelper(connectionId, new Message(Json.encode(eventBody.getBody())), handler);
    }

    protected static class MessageCallback implements com.microsoft.azure.sdk.iot.device.MessageCallback
    {
        ModuleClient _client;
        Handler<AsyncResult<EventBody>> _handler;
        String _inputName;

        public MessageCallback(ModuleClient client, String inputName, Handler<AsyncResult<EventBody>> handler)
        {
            this._client = client;
            this._inputName = inputName;
            this._handler = handler;
        }

        public synchronized IotHubMessageResult execute(Message msg, Object context)
        {
            System.out.println("MessageCallback called");
            this._client.setMessageCallback(this._inputName, null, null);
            String result = new String(msg.getBytes(), Message.DEFAULT_IOTHUB_MESSAGE_CHARSET);
            System.out.printf("result = %s%n", result);
            if (this._handler != null)
            {
                this._handler.handle(Future.succeededFuture(new EventBody(new JsonObject(result), new JsonObject(), new JsonObject())));
            }
            return IotHubMessageResult.COMPLETE;
        }
    }

    public void waitForInputMessage(String connectionId, String inputName, Handler<AsyncResult<EventBody>> handler)
    {
        System.out.printf("waitForInputMessage with %s, %s%n", connectionId, inputName);

        ModuleClient client = getClient(connectionId);
        if (client == null)
        {
            handler.handle(Future.failedFuture(new MainApiException(500, "invalid connection id")));
        }
        else
        {
            MessageCallback callback = new MessageCallback(client, inputName, handler);
            System.out.printf("calling setMessageCallback%n");
            client.setMessageCallback(inputName, callback, null);
        }
    }

    private static class DeviceMethodCallbackImpl implements MethodCallback
    {
        public Handler<AsyncResult<Void>> _handler;
        public String _requestBody;
        public String _responseBody;
        public String _methodName;
        public int _statusCode;
        public ModuleClient _client;

        public void reset()
        {
            this._methodName = null;
            this._handler = null;
        }

        @Override
        public DirectMethodResponse onMethodInvoked(String methodName, DirectMethodPayload methodData, Object context)
        {
            System.out.printf("method %s called%n", methodName);
            if (methodName.equals(this._methodName))
            {
                System.out.println("Received payload:");
                System.out.println(methodData.getPayloadAsJsonString());

                Object methodDataObject = methodData.getPayload(Map.class);
                System.out.printf("methodData: %s%n", methodDataObject);

                if (methodDataObject.equals(this._requestBody) ||
                    Json.encode(methodDataObject).equals(this._requestBody))
                {
                    System.out.printf("Method data looks correct.  Returning result: %s%n", _responseBody);
                    this._handler.handle(Future.succeededFuture());
                    this.reset();
                    return new DirectMethodResponse(this._statusCode, this._responseBody);
                }
                else
                {
                    System.out.printf("method data does not match.  Expected %s%n", this._requestBody);
                    this._handler.handle(Future.failedFuture("methodData does not match"));
                    this.reset();
                    return new DirectMethodResponse(500, "methodData not received as expected");
                }
            }
            else
            {
                this._handler.handle(Future.failedFuture("unexpected call: " + methodName));
                this.reset();
                return new DirectMethodResponse(404, "method " + methodName + " not handled");
            }
        }
    }

    DeviceMethodCallbackImpl _methodCallback = new DeviceMethodCallbackImpl();

    public void enableMethods(String connectionId, Handler<AsyncResult<Void>> handler)
    {
        ModuleClient client = getClient(connectionId);
        if (client == null)
        {
            handler.handle(Future.failedFuture(new MainApiException(500, "invalid connection id")));
        }
        else
        {
            IotHubEventCallbackImpl callback = new IotHubEventCallbackImpl();
            callback.setHandler(handler);
            client.subscribeToMethodsAsync(this._methodCallback, null, callback, null);
        }
    }

    public void WaitForMethodAndReturnResponse(String connectionId, String methodName, MethodRequestAndResponse requestAndResponse, Handler<AsyncResult<Void>> handler)
    {
        ModuleClient client = getClient(connectionId);
        if (client == null)
        {
            handler.handle(Future.failedFuture(new MainApiException(500, "invalid connection id")));
        }
        else
        {
            _methodCallback._handler = handler;
            _methodCallback._requestBody = Json.encode(((LinkedHashMap) requestAndResponse.getRequestPayload()).get("payload"));
            _methodCallback._responseBody = Json.encode(requestAndResponse.getResponsePayload());
            _methodCallback._statusCode = requestAndResponse.getStatusCode();
            _methodCallback._client = client;
            _methodCallback._methodName = methodName;
        }
    }

    private JsonObject makeMethodResultThatEncodesCorrectly(com.microsoft.azure.sdk.iot.device.edge.DirectMethodResponse result)
    {
        // Our JSON encoder doesn't like the way the MethodClass implements getPayload and getPayloadObject.  It
        // produces JSON that had both fields and the we want to return payloadObject, but we want to return it
        // in the field called "payload".  The easiest workaroudn is to make an empty JsonObject and copy the
        // values over manually.  I'm sure there's a better way, but this is test code.
        JsonObject fixedObject = new JsonObject();
        fixedObject.put("status", result.getStatus());

        // The payload type in DirectMethodRequest is a json string generated by Json.encode(),
        // so we get the payload as a json string as well on the response side.
        fixedObject.put("payload", result.getPayloadAsJsonString());
        return fixedObject;
    }


    public void invokeModuleMethod(String connectionId, String deviceId, String moduleId, MethodInvoke methodInvokeParameters, Handler<AsyncResult<Object>> handler)
    {
        ModuleClient client = getClient(connectionId);
        if (client == null)
        {
            handler.handle(Future.failedFuture(new MainApiException(500, "invalid connection id")));
        }
        else
        {
            DirectMethodRequest request = new DirectMethodRequest(methodInvokeParameters.getMethodName(), Json.encode(methodInvokeParameters.getPayload()), methodInvokeParameters.getResponseTimeoutInSeconds(), methodInvokeParameters.getConnectTimeoutInSeconds());
            try
            {
                com.microsoft.azure.sdk.iot.device.edge.DirectMethodResponse result = client.invokeMethod(deviceId, moduleId, request);
                handler.handle(Future.succeededFuture(makeMethodResultThatEncodesCorrectly(result)));
            } catch (Exception e)
            {
                handler.handle(Future.failedFuture(e));
            }
        }
    }

    public void sendOutputEvent(String connectionId, String outputName, EventBody eventBody, Handler<AsyncResult<Void>> handler)
    {
        System.out.printf("sendOutputEvent called for %s, %s%n", connectionId, outputName);
        System.out.println(eventBody);
        Message msg = new Message(Json.encode(eventBody.getBody()));
        msg.setOutputName(outputName);
        this.sendEventHelper(connectionId, msg, handler);
    }

    public void waitForDesiredPropertyPatch(String connectionId, Handler<AsyncResult<Twin>> handler)
    {
        System.out.printf("waitForDesiredPropertyPatch with %s%n", connectionId);

        ModuleClient client = getClient(connectionId);
        if (client == null)
        {
            handler.handle(Future.failedFuture(new MainApiException(500, "invalid connection id")));
        }
        else
        {
            this.setTwinHandler(res -> {
                if (res.succeeded())
                {
                    handler.handle(Future.succeededFuture(res.result()));
                }
                else
                {
                    handler.handle(res);
                }
            });

        }
    }

    public void getTwin(String connectionId, Handler<AsyncResult<Twin>> handler)
    {
        System.out.printf("getTwin with %s%n", connectionId);

        ModuleClient client = getClient(connectionId);
        if (client == null)
        {
            handler.handle(Future.failedFuture(new MainApiException(500, "invalid connection id")));
        }
        else
        {
            this.setTwinHandler(handler);
            try
            {
                com.microsoft.azure.sdk.iot.device.twin.Twin twin = client.getTwin();
                TwinCollection desiredProperties = twin.getDesiredProperties();
                for (String key : desiredProperties.keySet())
                {
                    onPropertyChanged(new Property(key, desiredProperties.get(key)), null);
                }
            } catch (Exception e)
            {
                this.setTwinHandler(null);
                handler.handle(Future.failedFuture(e));
            }
        }
    }

    public void sendTwinPatch(String connectionId, Twin twin, Handler<AsyncResult<Void>> handler)
    {
        System.out.printf("sendTwinPatch called for %s%n", connectionId);
        System.out.println(twin.toString());

        ModuleClient client = getClient(connectionId);
        if (client == null)
        {
            handler.handle(Future.failedFuture(new MainApiException(500, "invalid connection id")));
        }
        else
        {
            TwinCollection reportedProperties = new TwinCollection((LinkedHashMap<String, Object>)twin.getReported());
            System.out.println("reported properties size" + reportedProperties.size());
            System.out.println("reported properties:" + reportedProperties);
            this._deviceTwinStatusCallback.setHandler(handler);
            try
            {
                client.updateReportedProperties(reportedProperties);
                handler.handle(Future.succeededFuture());
            }
            catch (Exception e)
            {
                this._deviceTwinStatusCallback.setHandler(null);
                handler.handle(Future.failedFuture(e));
            }
        }
    }

    public void Cleanup()
    {
        Set<String> keys = this._map.keySet();
        if (!keys.isEmpty())
        {
            for (String key : keys)
            {
                this._closeConnection(key);
            }
        }
    }
}
